src/game.zig

```diff
@@ -567,13 +567,31 @@ pub const Player = struct { // MARK: Player
 				}
 			} else return;
 
-			// Check if there is already a slot with that item type
-			for(0..12) |slotIdx| {
+			for(0..(main.gui.windowlist.inventory.itemSlots.len + main.gui.windowlist.hotbar.itemSlots.len)) |slotIdx| {
 				if(std.meta.eql(inventory.getItem(slotIdx), item)) {
-					if(isCreative()) {
-						inventory.fillFromCreative(@intCast(slotIdx), item);
+					if(slotIdx < (main.gui.windowlist.hotbar.itemSlots.len)) {
+						// when item is in hotbar
+						selectedSlot = @intCast(slotIdx);
+					} else {
+						// when item is in inventory
+						const targetSlot = blk: {
+							if(inventory.getItem(selectedSlot) == null) break :blk selectedSlot;
+							// Look for an empty slot
+							for(0..main.gui.windowlist.hotbar.itemSlots.len) |slotId| {
+								if(inventory.getItem(slotId) == null) {
+									break :blk slotId;
+								}
+							}
+							break :blk selectedSlot;
+						};
+						var carried: Inventory = undefined;
+						carried = Inventory.init(main.globalAllocator, 1, .normal, .{.hand = main.game.Player.id}, .{});
+						defer carried.deinit(main.globalAllocator);
+
+						inventory.depositToAny(@intCast(slotIdx), carried, inventory.getAmount(slotIdx));
+						inventory.depositOrSwap(@intCast(targetSlot), carried);
+						inventory.depositOrSwap(@intCast(slotIdx), carried);
```

Comment:
This may cause unintended behavior if the player has something in the carried slot (this an edge-case that can happen if you load the game after storing it with something in your hand)

Ideally it should instead just swap the two slots directly.